include/pybind11/pytypes.h

```diff
@@ -1829,12 +1829,12 @@ class capsule : public object {
             // guard if destructor called while err indicator is set
             error_scope error_guard;
             auto destructor = reinterpret_cast<void (*)(void *)>(PyCapsule_GetContext(o));
-            if (PyErr_Occurred()) {
+            if (destructor == nullptr && PyErr_Occurred()) {
                 throw error_already_set();
             }
             const char *name = get_name_in_error_scope(o);
             void *ptr = PyCapsule_GetPointer(o, name);
-            if (ptr == nullptr) {
+            if (ptr == nullptr && PyErr_Occurred()) {
```

Comment:
I looked at this recently in connection with something completely different:

https://docs.python.org/3/c-api/capsule.html#c.PyCapsule_New

> Create a [PyCapsule](https://docs.python.org/3/c-api/capsule.html#c.PyCapsule) encapsulating the pointer. The pointer argument may not be NULL.

I found it impossible to create a capsule with a NULL pointer.

Therefore I think it's better to not change this line. If no error is set, something is very wrong, and it'll be better to let `error_already_set()` tell us:

https://github.com/pybind/pybind11/blob/da104a9efd0357b5c144c67eb641ae0b9e23012d/include/pybind11/pytypes.h#L451

Reply:
@rwgk Ah, you are right. I just updated this PR.